lib/Onyx.js

```diff
@@ -1085,41 +1086,48 @@ function set(key, value) {
         return Promise.resolve();
     }
 
-    const valueWithoutNull = removeNullValues(key, value);
-
-    if (valueWithoutNull === null) {
-        return Promise.resolve();
-    }
+    // If the value is null, we remove the key from storage
+    const {value: valueAfterRemoving, wasRemoved} = removeNullValues(key, value);
 
     if (hasPendingMergeForKey(key)) {
-        Logger.logAlert(`Onyx.set() called after Onyx.merge() for key: ${key}. It is recommended to use set() or merge() not both.`);
+        delete mergeQueue[key]
     }
 
-    const hasChanged = cache.hasValueChanged(key, valueWithoutNull);
+    const hasChanged = cache.hasValueChanged(key, valueAfterRemoving);
 
     // This approach prioritizes fast UI changes without waiting for data to be stored in device storage.
-    const updatePromise = broadcastUpdate(key, valueWithoutNull, hasChanged, 'set');
+    const updatePromise = broadcastUpdate(key, valueAfterRemoving, 'set', hasChanged, wasRemoved);
 
-    // If the value has not changed, calling Storage.setItem() would be redundant and a waste of performance, so return early instead.
-    if (!hasChanged) {
+    // If the value has not changed or the key got removed, calling Storage.setItem() would be redundant and a waste of performance, so return early instead.
+    if (!hasChanged || wasRemoved) {
         return updatePromise;
     }
 
-    return Storage.setItem(key, valueWithoutNull)
-        .catch((error) => evictStorageAndRetry(error, set, key, valueWithoutNull))
+    return Storage.setItem(key, valueAfterRemoving)
+        .catch((error) => evictStorageAndRetry(error, set, key, valueAfterRemoving))
         .then(() => updatePromise);
 }
 
 /**
  * Storage expects array like: [["@MyApp_user", value_1], ["@MyApp_key", value_2]]
  * This method transforms an object like {'@MyApp_user': myUserValue, '@MyApp_key': myKeyValue}
- * to an array of key-value pairs in the above format
+ * to an array of key-value pairs in the above format and removes key-value pairs that are being set to null
  * @private
  * @param {Record} data
- * @return {Array} an array of key - value pairs <[key, value]>
+ * @return {Array} an array of key - value pairs and a boolean which indicates if the key got removed <[key, value]>
```

Comment:
I don't see where we're returning the `wasRemoved` boolean.

Reply:
updated the comment! this was a relict from previous changes